driver/src/main/com/mongodb/client/gridfs/GridFSDownloadStreamImpl.java

```diff
@@ -0,0 +1,159 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.mongodb.client.gridfs;
+
+import com.mongodb.MongoGridFSException;
+import com.mongodb.client.MongoCollection;
+import com.mongodb.client.gridfs.model.GridFSFile;
+import org.bson.BsonDocument;
+import org.bson.BsonInt32;
+import org.bson.BsonValue;
+
+import java.io.IOException;
+
+import static com.mongodb.assertions.Assertions.notNull;
+import static java.lang.String.format;
+
+class GridFSDownloadStreamImpl extends GridFSDownloadStream {
+    private final GridFSFile fileInfo;
+    private final MongoCollection<BsonDocument> chunksCollection;
+    private final BsonValue fileId;
+    private final long length;
+    private final int chunkSizeInBytes;
+    private final int numberOfChunks;
+    private int chunkIndex;
+    private int bufferOffset;
+    private long currentPosition;
+    private byte[] buffer = null;
+    private boolean closed;
+
+    GridFSDownloadStreamImpl(final GridFSFile fileInfo, final MongoCollection<BsonDocument> chunksCollection) {
+        this.fileInfo = notNull("file information", fileInfo);
+        this.chunksCollection = notNull("chunks collection", chunksCollection);
+
+        fileId = fileInfo.getId();
+        length = fileInfo.getLength();
+        chunkSizeInBytes = fileInfo.getChunkSize();
+        numberOfChunks = (int) Math.ceil((double) length / chunkSizeInBytes);
+    }
+
+    @Override
+    public GridFSFile getGridFSFile() {
+        return fileInfo;
+    }
+
+    @Override
+    public int read() throws IOException {
+        byte[] b = new byte[1];
+        int res = read(b);
+        if (res < 0) {
+            return -1;
+        }
+        return b[0] & 0xFF;
+    }
+
+    @Override
+    public int read(final byte[] b, final int off, final int len) throws IOException {
+        checkClosed();
+
+        if (currentPosition == length) {
+            return -1;
+        } else if (buffer == null) {
+            buffer = getBuffer(chunkIndex);
+            bufferOffset = 0;
+        } else if (bufferOffset == buffer.length) {
+            chunkIndex += 1;
+            buffer = getBuffer(chunkIndex);
+            bufferOffset = 0;
+        }
+
+        int r = Math.min(len, buffer.length - bufferOffset);
+        System.arraycopy(buffer, bufferOffset, b, off, r);
+        bufferOffset += r;
+        currentPosition += r;
+        return r;
+    }
+
+    @Override
+    public long skip(final long bytesToSkip) throws IOException {
+        checkClosed();
+        if (bytesToSkip <= 0) {
+            return 0;
+        }
+
+        long skippedPosition = currentPosition + bytesToSkip;
+        bufferOffset = (int) skippedPosition % chunkSizeInBytes;
```

Comment:
If I comment out this line of code, "GridFSDownloadStreamSpecification#should skip to the correct point" still passes, so there is a missing assertion about what read does after a skip.